Repair interrupted state store migrations (#224)
microsoft/dcp: internal/statestore/migrations/README.md

2. Advances `schema_migrations` when moving to the next major version.
3. Applies the new major version's minor migrations.

### Serializing concurrent migrations

`golang-migrate` does not serialize migrations across processes when using SQLite. Its `Driver` interface lets an implementation opt out of locking, and the SQLite driver does: `Lock` only flips an in-process flag on the driver instance. DCP builds a fresh migration runner for every `Open`, so that flag serializes nothing at all.

> The lock DCP takes in `migrate` is therefore the only thing that keeps two DCP instances from migrating the same store at once, and it must stay wrapped around the whole read, repair, and migrate sequence. `golang-migrate` reads the current version and then writes the version marker in separate transactions, so instances that are not serialized can both read the same version, both decide to migrate, and both run the same migration — the loser fails to start or leaves a dirty marker behind.

Reviewer: ```suggestion
The lock DCP takes in `migrate` is therefore the only thing that keeps two DCP instances from migrating the same store at once, and it must stay wrapped around the whole read schema version and migrate sequence.
```

> The lock is an advisory file lock held on an open file descriptor, so the operating system releases it when the owning process exits, including a crash. DCP waits for the lock rather than failing when another instance holds it, which also covers Windows releasing the lock asynchronously after a process dies. An interrupted migration therefore leaves a dirty version marker but never a permanently stuck lock, which is what lets the next DCP acquire the lock and repair the marker.

Reviewer: ```suggestion
The lock is an advisory file lock held on an open file descriptor, so the operating system releases it when the owning process exits, including a crash. DCP waits for the lock rather than failing when another instance holds it. An interrupted migration therefore leaves a dirty version marker but never a permanently stuck lock, which is what lets the next DCP acquire the lock and clear the marker.
```

> ### Recovering a dirty version

Reviewer: ```suggestion
### How DCP handles failed migrations
```

> `golang-migrate` marks a version as dirty before running its SQL and clears the dirty flag after the SQL succeeds. A dirty version therefore means a migration did not complete. That happens when DCP is interrupted part way through a migration, and also when an older DCP re-runs a migration a newer DCP already applied: the older binary expects a version marker the newer layout no longer records, re-runs the migration, and its SQL fails because the schema change is already present.

Reviewer: ```suggestion
`golang-migrate` marks a version as dirty before running its SQL and clears the dirty flag after the SQL succeeds. A dirty version therefore means a migration did not complete. That can happen when DCP is interrupted part way through a migration.
```

Reviewer: Here I assume it never really happens that DCP tries to apply the same migration that already succeeded previously (because of probing described below). So the whole second part of this paragraph seems incorrect.

Author: Going forward that will be the case; the old wording was overly focused on the specific conditions that led to the major/minor split in the first place (and could be a source of state store corruption). Reworded to link back to the section on that issue instead of calling it out as an ongoing concern.

> DCP repairs a dirty version instead of failing startup. Every migration registers an *applied probe* in `schema.go`: a SQL predicate that reports whether that migration's schema changes are present. DCP evaluates the probe for the dirty version and rewrites the version table:

Reviewer: ```suggestion
Every migration registers an *applied probe* in `schema.go`: a SQL predicate that reports whether that migration's schema changes are present. DCP evaluates the probe for the dirty version(s) and rewrites the version table:
```

> - The probe reports true, so the migration committed. DCP records the dirty version as clean and continues with the following migration.

Reviewer: ```suggestion
- If the probe reports true, the migration committed. DCP clears the dirty flag and continues with the following migration.
```

> - The probe reports false, so the migration never committed. DCP records the preceding version as clean, or empties the stream when the dirty version was the first migration, and the migration runs again.

Reviewer: ```suggestion
- If probe reports false, the migration never committed. DCP marks the preceding version as clean (or assumes no migrations happened if the dirty version was the first migration), clears the dirty flag for the failed migration, and re-applies it.
```

DCP does not repair every dirty marker:

- A dirty minor version newer than any migration embedded in this binary cannot be probed, because the migration belongs to a newer DCP. Minor migrations are additive, so every migration this binary needs is present regardless of whether the interrupted one committed; DCP logs the condition, leaves the marker alone, and lets the binary that owns the migration repair it.

> - An unknown major version fails startup whether it is dirty or clean. Major versions are reserved for changes that older binaries cannot use safely.

Reviewer: ```suggestion

A dirty minor version newer than any migration embedded in current binary cannot be probed--it was attempted by a newer DCP. Minor migrations are additive, so every migration this binary needs is present regardless of whether the interrupted one committed. DCP logs the condition, leaves the dirty marker alone, and lets the binary that owns the migration repair it.

An unknown major version fails startup whether it is dirty or clean. Major versions are reserved for changes that older binaries cannot use safely.
```

> ### Why repairing a dirty version is safe

Reviewer: ```suggestion
### Migrations are atomic
```

A dirty marker for version *v* has exactly two possible meanings:

1. The migration's transaction never committed, so the schema is exactly at version *v-1*.

> 2. The transaction committed but DCP stopped before clearing the dirty flag, so the schema is exactly at version *v*.

Reviewer: ```suggestion
Every DCP minor schema migration is additive and atomic because it runs in a single transaction. Only two outcomes are possible:

1. The migration's transaction failed to commit, so the schema is exactly at version *v-1*.
2. The transaction committed so the schema is exactly at version *v*.
```

Author: Left both words out on purpose.

Atomicity comes from the driver's transaction wrapper, which applies to every migration regardless of which stream it belongs to — `TestMigrationsAreTransactional` checks every embedded file, major and minor. Scoping the claim to minor migrations would imply a major migration can be left half-applied, which contradicts the last paragraph of this section ("DCP detects existing schema version for major and minor migrations in the same way").

"additive" is a compatibility constraint rather than an atomicity one. The two-outcome result follows from the transaction alone; a future major migration that drops a column would still be atomic, just not compatible. Additivity is already covered under the minor migration rules below, where it is actually load-bearing.

Reviewer: OK so then say "Every DCP schema migration is atomic because..." (rest unchanged)

Author: That is what it already says — line 45 reads "Every DCP schema migration is atomic because it runs in a single transaction." I only dropped the two words from your suggestion, the rest was unchanged. Resolving.

> There is no third case where the schema sits part way between the two. That is what makes the probe result a reliable answer rather than a guess, and it holds because of two constraints this repository places on migration authoring, both enforced by unit tests:

Reviewer: ```suggestion
There is no third case where the schema sits part way between the two. This means that the dirty flag used by `golang-migrate` is irrelevant--it is updated outside of the main migration transaction, and can remain set even if the migration succeeds. This is why DCP clears it automatically in most cases, as described above.

The use of single transaction for each migration is also what makes the probe result a reliable answer rather than a guess, and it holds because of two constraints this repository places on migration authoring, both enforced by unit tests:
```

Author: Took the "no third case" sentence and the "updated outside the migration transaction" clause, but went with "not authoritative on its own" instead of "irrelevant".

The flag is what triggers repair in the first place. A clean flag means DCP skips probing entirely and trusts the recorded version, so the flag is doing real work — it just is not authoritative about what the schema actually contains. Calling it irrelevant would suggest we could ignore it, and then there would be nothing left to key repair off of.

Reviewer: Sounds good, thanks

- **Migration SQL never commits implicitly.** The SQLite migration driver wraps each migration in a single transaction, so `BEGIN`, `COMMIT`, `END`, `ROLLBACK`, `VACUUM`, `PRAGMA`, `ATTACH`, and `DETACH` are rejected. `TestMigrationsAreTransactional` checks every embedded migration file. SQLite treats `END` as an alias for `COMMIT`, so only an `END` that closes a `CASE` expression is allowed. A migration that ended its transaction part way through would leave a schema that is neither version, and no probe could describe it.

> - **Every migration has an applied probe that discriminates.** The probe must report false before the migration runs and true afterwards. Without a probe DCP cannot tell which of the two cases it is in, and startup fails with the dirty version. A probe that does not distinguish the two states is worse: DCP would keep a version marker for a migration that never ran, permanently skipping it.

Reviewer: ```suggestion
- **Every migration has an applied probe that discriminates.** The probe must reliably report `false` before the migration is applied and `true` afterwards. 
```

> Because the schema always matches one of the two versions exactly, DCP can determine the real migration state and record it. Repair only rewrites the version marker; it never edits schema or data. When the probe reports false the schema is exactly at *v-1*, so re-running the migration is an ordinary first application rather than a retry against a half-changed schema. DCP repairs the major version stream and each major version's minor stream the same way.

Reviewer: ```suggestion
Because the schema always matches one of the two versions exactly, DCP can determine the real migration state and record it. When the probe reports false the schema is exactly at *v-1*, so re-running the migration is an ordinary first application rather than a retry against a half-changed schema. DCP detects existing schema version for major and minor migrations in the same way, and clears the `golang-migrate` dirty flag for both.
```

Author: Took the rest of this, but kept "Repair only rewrites the version marker; it never edits schema or data."

It bounds what repair actually touches — one row in the version table — which is the first thing I would want to know as someone reading that DCP automatically recovers a state store it found in a bad state. It is also what justifies keeping repair inside the migration lock in [Serializing concurrent migrations](https://github.com/microsoft/dcp/blob/danegsta-silver-fiesta/internal/statestore/migrations/README.md#serializing-concurrent-migrations): repair is a write, not just a probe evaluation, so it cannot be hoisted out.

That said, this is the least load-bearing of the three and I will drop it if you still think it is stating the obvious.

Reviewer: I probably should be more clear: I do not like the term "repair" here. We are not really fixing anything; we are just determining what the state of the schema is and (for migrations that the current instance of DCP knows about) clearing the dirty flag. 

Hate to say this, but "reconcile" (as in "reconcile version table") would be a good word choice, but let's not use this since big part of DCP is a bunch of contorllers. Maybe "version table update" is a good-enough compromise.

Reviewer: ... but I am not going to ask you to change it everywhere. Will just point out instances where I think the term is ambiguous enough to hurt.

Author: That reframing is right and it is worth changing everywhere, so I did — `511b382` renames repair to recovery across the package, the tests, and the README. `repairDirtySchemaVersion` is now `recoverDirtySchemaVersion`, the log message is "Recovered an interrupted state store migration", and the tests are `TestOpenRecovers...`.

Agreed on avoiding "reconcile" given what that word already means in this codebase. Where the sentence is about the marker specifically rather than the operation as a whole I went with "clear" instead, since that is literally what happens: "DCP does not clear every dirty marker", "lets the binary that owns the migration clear it".

I did keep the sentence itself, now reading "Recovery only rewrites the version marker; it never edits schema or data." Reading your two replies together, the objection is the term rather than the claim, and once the term stops implying damage the sentence is just bounding what the operation touches. Still happy to cut it if you disagree.

- Do not add unique indexes, check constraints, foreign keys, changed primary keys, or other constraints that can reject writes accepted by older binaries without an explicit compatibility design.
- Preserve `resource_locks` so old and new DCP processes coordinate through the same lock records.
- Preserve persistent resource tables and records so older binaries can continue reading and updating them.

> - Do not include `BEGIN`, `COMMIT`, `END`, `ROLLBACK`, `VACUUM`, `PRAGMA`, `ATTACH`, or `DETACH`. These end the transaction the driver wraps the migration in, and [Why repairing a dirty version is safe](#why-repairing-a-dirty-version-is-safe) explains why that would make a dirty version unrecoverable. Only an `END` that closes a `CASE` expression is allowed.

Reviewer: ```suggestion
- Do not include `BEGIN`, `COMMIT`, `END`, `ROLLBACK`, `VACUUM`, `PRAGMA`, `ATTACH`, or `DETACH`. These end the transaction the driver wraps the migration in, and [Migrations are atomic](#migrations-are-atomic) explains why that would make a failed migration potentially unrecoverable. Only an `END` that closes a `CASE` expression is allowed.
```
